### ISSUE TITLE
Transitions can not have attributes that are `select(`'d

### ISSUE THREAD

**Issue by Mythra** [bazelbuild/bazel#12386]
### Description of the problem:

When attempting to use a `select` for an attribute that is a transition Bazel ignores these attributes.

_NOTE: This seems to be known, and have a buganizer id: https://github.com/bazelbuild/bazel/blob/master/src/main/java/com/google/devtools/build/lib/analysis/starlark/StarlarkRuleTransitionProvider.java#L88-L90 , but I couldn't find a public issue that matches it_

_NOTE: https://github.com/bazelbuild/bazel/blob/master/src/main/java/com/google/devtools/build/lib/analysis/starlark/StarlarkRuleTransitionProvider.java#L46-L48 this seems to be because of these corner cases (perhaps there's more info on the buganizer as to what those are?)_

### Bugs: what's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

I've run into this issue trying to map bazel buildsettings into `gotags`: https://github.com/Mythra/rules_go_repro provides a minimal example actually using rules_go (the exact case I'm trying to support). Specifically you can look at the macro here: https://github.com/Mythra/rules_go_repro/blob/trunk/pkg/macro.bzl#L4-L18

### What operating system are you running Bazel on?

Arch Linux

### What's the output of `bazel info release`?

release 3.6.0

### Any other information, logs, or outputs that you want to share?

I'd be curious in knowing what these corner cases were, and if you can think of any other alternatives we can use while waiting for transitions to minimally support select, so users can specify arguments to control the build, and have them populate to gotags in the transition rule.

**Comment by gregestren** [comment]
Not really much more context in the Buganizer bug, but https://docs.google.com/document/d/1MDbxk21nti40uoMzt7h02213086KtEb2T_WEKVmMjuQ/ should have some more.

I remember there being some real technical subtlety here. Assigning to @juliexxia for more recollection.

**Comment by juliexxia** [comment]
Thanks for linking the doc, Greg! Read about the corner case in this section with a linked figure for easier visualization: https://docs.google.com/document/d/1MDbxk21nti40uoMzt7h02213086KtEb2T_WEKVmMjuQ/edit#bookmark=id.uk5qj36g606m

Note that these restrictions are only for transitions directly on rules, not transitions on attributes so if you can shift the transition onto an attribute you can fully get around them.

**Comment by Mythra** [comment]
Hey @juliexxia I apprecaite the doc with additional info, but I'm curious what would this transition on an attribute look like? (Also perhaps this example would be useful in the doc as a workaround or "the current state"). I didn't even realize transitions could be on attributes. I guess I'm confused on how the current rule setup I'm working with is:

```starlark
go_test(
    <snipped>
    gotags = select({
      '//buildsettings:sql_driver_sqlite': ['sqlite'],
      '//buildsettings:sql_driver_postgres': ['postgres'],
    }),
    <snipped>
  )
```

which ends up going to a transition on a rule. Could I introduce a transition on the gotags attribute outside of rules_go (where go_test is defined) to get this to work, what would that look like if so? Or would I need to change the actual go_test transition rule? If I need to change the go_test transition rule what would I be looking at for taking an attribute of a rule, and making it a transition on just that attribute (looking for a simple example like the one we have in that doc)?

Sorry if these are obvious questions but the docs for transitions are not clear to me on any of it's functionality :sweat_smile:

**Comment by juliexxia** [comment]
Have you seen the general starlark transitions documentation? Probably a helpful place to start: https://docs.bazel.build/versions/master/skylark/config.html

> Could I introduce a transition on the gotags attribute outside of rules_go (where go_test is defined) to get this to work, what would that look like if so? Or would I need to change the actual go_test transition rule? 

Right, transitions are defined on rules, not targets. 

> If I need to change the go_test transition rule what would I be looking at for taking an attribute of a rule, and making it a transition on just that attribute (looking for a simple example like the one we have in that doc)?

Here's the section on attaching transitions to the rule (incoming transition, rule applies to itself (and all its dependencies)) or the attribute (outgoing transition, rule applies to some of its dependencies): https://docs.bazel.build/versions/master/skylark/config.html#attaching-transitions There's a link to an end to end example which might be helpful

> Sorry if these are obvious questions but the docs for transitions are not clear to me on any of it's functionality 😅

No apologies needed, let me know if there's anything still lacking in the SBC docs

**Comment by ZackButcher** [comment]
Hey @juliexxia, thanks for the comments - I'm working with @Mythra on this. Honestly the transitions docs are a bit hard to follow as someone not deeply familiar with Bazel internals. However, I feel like I'm right on the edge of understanding them. Two things I think might help improve the docs:
1. In the [`Defining`](https://docs.bazel.build/versions/master/skylark/config.html#defining) sections, y'all declare:
    ```bazel
    # example/transitions/transitions.bzl
    def _impl(settings, attr):
        _ignore = (settings, attr)
        return {"//example:favorite_flavor" : "MINT"}
    
    hot_chocolate_transition = transition(
        implementation = _impl,
        inputs = [],
        outputs = ["//example:favorite_flavor"]
    )

    # separately
    
    # example/transitions/transitions.bzl
    def _impl(settings, attr):
        _ignore = (settings, attr)
        return [
            {"//example:favorite_flavor" : "LATTE"},
            {"//example:favorite_flavor" : "MOCHA"},
        ]
    
    coffee_transition = transition(
        implementation = _impl,
        inputs = [],
        outputs = ["//example:favorite_flavor"]
    )
    ```

    Then you give partial examples:

    ```bazel
    # example/rules.bzl
    load("example/transitions:transitions.bzl", "hot_chocolate_transition")
    drink_rule = rule(
        implementation = _impl,
        cfg = hot_chocolate_transition,
        ...
    
    # separately
    
    # example/rules.bzl
    load("example/transitions:transitions.bzl", "coffee_transition")
    drink_rule = rule(
        implementation = _impl,
        attrs = { "dep": attr.label(cfg = coffee_transition)}
        ...
    ```
    It would be incredibly helpful if the docs showed the attributes on an instance of `drink_rule` with and without (or before and after) the transition in each instance. And better, include the `bazel query` command I could run to see it myself (I have a feeling I could get bazel to spit it out, but I don't know how).
1. It would be awesome to have a diagram (and some examples) of Bazel's internal execution order e.g. under a `bazel build` execution. This would help solidify how settings are applied via transitions on rules to compute targets (I _think_ that's close to the correct relationships? That's what I'm to get at). I'm sure this is on the site somewhere and I just missed it.

---

In our case specific case, I _think_ you're saying to attempt something along the lines of (this is not totally correct Skylark, just trying to rough out the idea):

```bazel
string_flag(
    name = 'sql_driver',
    values = ['sqlite', 'postgres'],
    build_setting_default = 'sqlite',
)

config_setting(
    name = 'sql_driver_sqlite',
    flag_values = {
      ':sql_driver': 'sqlite',
    }
)

config_setting(
    name = 'sql_driver_postgres',
    flag_values = {
      ':sql_driver': 'postgres',
    }
)


drivers = ['postgres', 'sqlite']
def _sql_gotags_transition(settings, attr):
    val = settings["//:sql_driver"]
    if val == "postgres":
      return {"gotags": ["bazel", "postgres"] }
    elif val == "sqlite":
      return {"gotags": ["bazel", "sqlite"] }
    else
      fail(":sql_driver has value " + val + " not in allowed values: [" + ", ".join(drivers) + "]")

add_sql_gotags = transition(
    implementation = _sql_gotags_transition,
    inputs = ["//:sql_driver"]
    outputs = ["gotags"]
)

go_binary_with_tags = rule(
    implementation = go_binary,
    attrs = { "gotags" = attr.label(cfg = add_sql_gotags) }
) 
```

Is that a good path to pursue, or is there a different way we should attempt this?

**Comment by juliexxia** [comment]
Thanks for the doc feedback! 

> It would be incredibly helpful if the docs showed the attributes on an instance of drink_rule with and without (or before and after) the transition in each instance. And better, include the bazel query command I could run to see it myself (I have a feeling I could get bazel to spit it out, but I don't know how).

When you say attributes, I assume you mean settings like `//example:favorite_flavor` here? The term attributes in blaze generally refer to the things defined by the `attr` parameter of a `rule()` function. Transitions affect build settings (traditionally known as flags) which are parts of the configuration, not attributes on a target (target == instance of a rule). If I'm assuming right, did you get a chance to look at our end-to-end examples? They're linked in the sections: https://github.com/bazelbuild/examples/tree/master/rules/starlark_configurations/attaching_transitions_to_rules (cd up to see the rest of them)

By the way,`bazel query` operates after the loading phase (see below) and prints out targets, before transitions trigger as part of configuration creation. `bazel cquery` operates after the analysis phase and prints out configured targets, after transitions trigger. Here are the cquery docs, particularly pointed at --transitions which might be an output format that is helpful for you here: https://docs.bazel.build/versions/master/cquery.html#transitions

> It would be awesome to have a diagram (and some examples) of Bazel's internal execution order e.g. under a bazel build execution. This would help solidify how settings are applied via transitions on rules to compute targets (I think that's close to the correct relationships? That's what I'm trying to get at). I'm sure this is on the site somewhere and I just missed it.

Blaze has three phases: loading, analysis, and execution. Here's the part of our docs that discusses the different phases of a blaze build: https://docs.bazel.build/versions/master/guide.html#phases-of-a-build Would it be helpful to link them as suggested pre-reading at the top of the page? The TL;DR (but I recommend R)- The loading phase reads BUILD files and loads targets, like //foo. The analysis phase takes these targets and turns them into configured targets, e.g. //foo with --cpu=k8 and //foo with --cpu=arm for how a single target can turn into multiple configured targets. Transitions are applied during the analysis phase to create these configurations. The execution phase executes all the actions defined in the rules implementation functions.

**Comment by juliexxia** [comment]
From your example of what you want to do, I think there's some confusion on what transitions can read and set. Transitions can read attributes and settings but can only set setting. Attribute values cannot be set during transitions. Why are you trying to change an attribute? Can you store the information inside a build setting? Are you hoping to read in inside the rule's implementation function? You can read build settings' values inside rule implementation functions  by declaring that the rule depends on the build setting. e2e example: https://github.com/bazelbuild/examples/tree/master/rules/starlark_configurations/use_skylib_build_setting

**Comment by Mythra** [comment]
Hey @juliexxia ,

The implementation above was based on us trying to understand transitions, but admittedly it's very hard to parse for us. Sounds like we were on the wrong track. To recap everything we know/what we're trying to do, we have a series of build_settings, to make it clearer I'll use only the one we've used before a `sql_driver` flag :

```starlark
load("@bazel_skylib//rules:common_settings.bzl", "string_flag")

string_flag(
  name = 'sql_driver',
  values = ['sqlite', 'postgres'],
  build_setting_default = 'sqlite',
)
config_setting(
  name = 'sql_driver_sqlite',
  flag_values = {
    ':sql_driver': 'sqlite',
  }
)
config_setting(
  name = 'sql_driver_postgres',
  flag_values = {
    ':sql_driver': 'postgres',
  }
)
```

We want to "read" this build-setting, to have it determine what dependencies to pull in, what "buildtags" to set, and whether or not to use cgo (as sqlite requires cgo, but nothing else does) when building the project with, ultimately what we want is something like the following:

```starlark
go_test(
      name = 'go_test',
      importpath = 'github.com/myorg/myrepo/package_test',
      srcs = glob(["**/*_test.go"]),
      embed = select({
        '//bazel/buildsettings:sql_driver_sqlite': ['@tetrate_resolve_deps//' + native.package_name() + ':sql_driver_sqlite_deps'],
        '//bazel/buildsettings:sql_driver_postgres': ['@tetrate_resolve_deps//' + native.package_name() + ':sql_driver_postgres_deps'],
      }),
      cgo = select({
        '//bazel/buildsettings:sql_driver_sqlite': True,
        '//conditions:default': False,
      }),
      pure = select({
        '//bazel/buildsettings:sql_driver_sqlite': "off",
        '//conditions:default': "on",
      }),
      gotags = ['bazel'] + select({
        '//bazel/buildsettings:sql_driver_sqlite': ['sqlite'],
        '//bazel/buildsettings:sql_driver_postgres': ['postgres'],
      }),
      static = 'on',
      size = test_size,
      x_defs = {'build': '{STABLE_VERSION_STRING}'},
)
```

These are all reads afaik, no writes. We were told: [Here](https://github.com/bazelbuild/rules_go/issues/2694#issuecomment-718803125), that the transition rule can't read the result of the select:

```
Unfortunately, I don't have good news for you. It looks like Bazel Starlark transitions can't read attributes that were set with select. See StarlarkRuleTransitionProvider.java.

There's a transition implementation function in Starlark that says which settings to change, based on attributes like gotags.

If that function can't access attributes set using select, then it can't completely perform the transition. The attribute appears not to exist, so the transition function ignores it.
```

(I'll also CC: @jayconrod who told us this who can probably mention more about what rules_go is doing incase you have questions about the actual transition/rule implemented there) Which lead to us opening this issue. You then mentioned:

```
Note that these restrictions are only for transitions directly on rules, not transitions on attributes so if you can shift the transition onto an attribute you can fully get around them.
```

This lead to us thinking it was something like what @ZackButcher mentioned above where we create a transition on the attribute itself. But now you mention:

```
Attribute values cannot be set during transitions. Why are you trying to change an attribute?
```

and now I'm confused left, right, and center 😓 . Do we want to do something with the attribute, and a transition? Do we want to modify the rule that is past the transition (effectively patching rules_go)? Do we want to do something else? I'm sorry if these are daft questions 😓 this just is very confusing to me.

**Comment by juliexxia** [comment]
It can all be very confusing! No apologies needed. Would y'all want to set up some time to meet over video? Might be faster than going back and forth on this github issue. I think some jargon is being lost in translation. 

If you're interested i can set some time up for us and a call. I'm working close to ET hours (I'm actually MT) and can meet today or tomorrow anytime other than 1pm ET. Let me know.

**Comment by Mythra** [comment]
Sure, happy to do that (also I'm MT too! :tada: ), Sounds like tomorrow around Noon ET would work for you? If so which email should I invite? :smile:

**Comment by juliexxia** [comment]
Let's just discuss in MT then (: Recently moved and still adjusting to the time math. 

Apologies I'm actually busy tomorrow until 11:30 MT (1:30 ET) so anytime after that would work for me.

**Comment by Mythra** [comment]
No worries. Wanna go for 3:30pm MT?

**Comment by juliexxia** [comment]
Sounds good!

**Comment by juliexxia** [comment]
As discussed just now - 

Here's the docs on the alias rule: https://docs.bazel.build/versions/master/be/general.html#alias

and an example of using it with a select: https://cs.opensource.google/bazel/bazel/+/master:tools/jdk/BUILD;l=290?q=%22actual%20%3D%20select(%22%20f:BUILD&ss=bazel

**Comment by juliexxia** [comment]
(Context: I'm moving on from Bazel on friday and spending some time triaging all bugs assigned to me before unassigning myself)

If I remember correctly, we found a suitable workaround for y'all on the call and I'm hoping since I haven't heard back it's doing it's job. Closing.
